dex/ws/wslink.go

```diff
@@ -70,28 +70,51 @@ type WSLink struct {
 	// Messages to the peer are routed through the outChan. This ensures
 	// messages are sent in the correct order, and satisfies the thread-safety
 	// requirements of the (*websocket.Conn).WriteMessage.
-	outChan chan []byte
+	outChan chan *sendData
 	// A master message handler.
 	handler func(*msgjson.Message) *msgjson.Error
 	// pingPeriod is how often to ping the peer.
 	pingPeriod time.Duration
 }
 
+type sendData struct {
+	data []byte
+	ret  chan<- error
+}
+
 // NewWSLink is a constructor for a new WSLink.
 func NewWSLink(addr string, conn Connection, pingPeriod time.Duration, handler func(*msgjson.Message) *msgjson.Error) *WSLink {
 	return &WSLink{
 		on:         true,
 		ip:         addr,
 		conn:       conn,
 		quit:       make(chan struct{}),
-		outChan:    make(chan []byte, outBufferSize),
+		outChan:    make(chan *sendData, outBufferSize),
 		pingPeriod: pingPeriod,
 		handler:    handler,
 	}
 }
 
-// Send sends the passed Message to the websocket peer.
+// Send sends the passed Message to the websocket peer. The actual writing of
+// the message on the peer's link occurs asynchronously. As such, a nil error
+// only indicates that the link is believed to be up and the message was
+// successfully marshalled.
 func (c *WSLink) Send(msg *msgjson.Message) error {
+	return c.send(msg, nil)
+}
+
+// SendNow is like send, but it waits for the message to be written on the
+// peer's link, returning any error from the write.
+func (c *WSLink) SendNow(msg *msgjson.Message) error {
+	writeErrChan := make(chan error, 1)
+	if err := c.send(msg, writeErrChan); err != nil {
+		return err
+	}
+	return <-writeErrChan
+}
+
+// CONCEPT
+func (c *WSLink) send(msg *msgjson.Message, writeErr chan<- error) error {
```

Comment:
There are paths where blocking is acceptable (the client's own websocket handler goroutines) and paths where it is not (requests from swapper, etc.).
I'm not using `SendNow`, but it's an idea that might improve things that presently have to wait for expire even though a write error is hit immediately.